### ISSUE TITLE
Stale sockets in CLOSE_WAIT state after b4899b952

### ISSUE THREAD

**Issue by vjt** [typhoeus/ethon#194]
Hello!

we bisected down that commit b4899b952 introduced in #136 to cause our application to keep sockets in `CLOSE_WAIT` after a transfer has ended. Reverting just that commit on the latest ethon makes the problem disappear.

As per the TCP spec, such a socket on our side indicates that our side has initiated closing a connection and it is waiting for an ACK from the peer. The ACK is seen reaching the host on which our system runs, but it appears that with b4899b952 ethon is not running `easy_cleanup` to process it and finalise the closure of the socket.

We experience this behaviour with ruby 2.6.5 and 2.7.3 and with both ffi 1.15.1 and 1.13.1, all leading to the same behaviour.

By just looking at the code I do not see anything blatantly wrong with it, but I am not an expert with FFI and its AutoPointer. I have a gut feeling of a deadlock somewhere - but before we proceed in more extensive analyses, does this ring any bell to you?

Thank you!

cc: @tagliala

**Comment by tagliala** [comment]
Hi,

we eventually came up with a reduced test case!

This issue is probably affecting everyone using ethon 0.14.0

**TL;DR** The code in the mutex is not executed when the cleanup is triggered by the Garbage Collector

### Requirements

1. Change the `TEST_URL`, or `python3 -m http.server 3000`
2. Copy the test case and run `RUBY_GC_MALLOC_LIMIT=10000 RUBY_GC_MALLOC_LIMIT_MAX=10000 ./ethon_194.rb`

### Reduced Test Case (ethon_194.rb)

```rb
#!/usr/bin/env ruby

# frozen_string_literal: true

ETHON_VERSION = '0.14.0'
TEST_URL = 'http://localhost:3000'

require 'bundler/inline'

begin
  puts "Ruby version: #{RUBY_VERSION}\n\n"

  gemfile(true) do
    source 'https://rubygems.org'

    gem 'ethon', ETHON_VERSION
  end
rescue Gem::LoadError => e
  puts "\nMissing Dependency:\n#{e.backtrace.first} #{e.message}"
rescue LoadError => e
  puts "\nError:\n#{e.backtrace.first} #{e.message}"
  puts DATA.read
  exit 1
end

require 'ethon'

# Monkey patching to provide debugging information
module Ethon
  if defined?(Ethon::Easy::Operations::PointerHelper)
    class Easy
      module Operations
        class PointerHelper
          class<<self
            alias original_release release

            def release(pointer)
              puts "*** PointerHelper.release called with #{pointer}"
              original_release(pointer)
            end
          end
        end
      end
    end
  end

  module Curl
    module_function

    def easy_cleanup(pointer)
      puts "***** Curl.easy_cleanup called with #{pointer}"
    end
  end
end

20.times do
  easy = Ethon::Easy.new
  easy.http_request(TEST_URL, :get)
  easy.perform
end
```

### Expected output

```
*** PointerHelper.release called with #<FFI::Pointer address=0x00007fdaec0a8000>
***** Curl.easy_cleanup called with #<FFI::Pointer address=0x00007fdaec0a8000>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80be971800>
***** Curl.easy_cleanup called with #<FFI::Pointer address=0x00007f80be971800>
[...]
```

### Current output

```
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80be971800>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b6c55400>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b6b99800>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80be420000>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80be958000>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b2829c00>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b28e6000>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b6c2dc00>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b6ba0000>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b6bb0000>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b6ba8000>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b37c9800>
*** PointerHelper.release called with #<FFI::Pointer address=0x00007f80b37c1800>
***** Curl.easy_cleanup called with #<FFI::Pointer address=0x00007f80b37c1800> 
```

**Comment by lloydwatkin** [comment]
Hi, is there any plans to get a release out with this in the near future? Since https://github.com/rubymem/ruby-mem-advisory-db/pull/29 this issue now causes our tests to fail. Thanks!
